atc/metric/metrics.go

```diff
@@ -675,20 +675,20 @@ func (event WorkersState) Emit(logger lager.Logger) {
 	for _, workerState := range event.WorkerStateByName {
 		_, exists := perStateCounter[workerState]
 		if !exists {
-			perStateCounter[workerState] = 0
+			perStateCounter[workerState] = 1
 			continue
 		}
 
 		perStateCounter[workerState] += 1
 	}
 
-	for state, count := range perStateCounter {
+	for _, state := range db.AllWorkerStates() {
+		count := perStateCounter[state]
 		if state == db.WorkerStateStalled && count > 0 {
```

Comment:
just wondering in this case if count can be < 0? does it make sense to check this count, since if we set `perStateCounter[workerState] = 1` to 1 means at least one will be in the map? correct me if I'm wrong?

Reply:
@kirillbilchenko I feel confident saying the count cannot be < 0, since the loop above only ever inserts values that are > 0. However, the value can be == 0.

This is because we are now iterating over `db.AllWorkerStates()` and it is quite possible (indeed, expected) that there will be some worker states that do not appear in the DB at the moment this metric is emitted (states like `landing` and `retiring` are intended to be ephemeral anyway and only occur when an operator intervenes).

From https://tour.golang.org/moretypes/22,
> If `key` is not in the map, then `elem` is the zero value for the map's element type.

Thus we can expect `count` to be zero in such cases. The other branch of this conditional is exercised in [this test case](https://github.com/concourse/concourse/pull/5015/files#diff-803b92969cc59c28c60a150bdc3c1b8fR55-R62), validating the need for that check.